genie-web/src/main/java/com/netflix/genie/web/resources/agent/AgentFileResourceImpl.java

```diff
@@ -0,0 +1,236 @@
+     *
+     * @return a {@link AgentFileStreamService.AgentFileResource}
+     */
+    public static AgentFileResourceImpl forNonExistingResource() {
+
+        final String description = AgentFileResourceImpl.class.getSimpleName()
+            + " [ non-existent ]";
+
+        return new AgentFileResourceImpl(
+            false,
+            null,
+            -1,
+            -1,
+            null,
+            description,
+            null
+        );
+    }
+
+
+    /**
+     * Factory method to create a resource for a remote file.
+     *
+     * @param uri              the resource URI
+     * @param size             the size of the file, as per latest manifest
+     * @param lastModifiedTime the last modification time, as per latest manifest
+     * @param relativePath     the path of the file relative to the root of the job directory
+     * @param jobId            the id of the job this file belongs to
+     * @param inputStream      the input stream to read this file content
+     * @return a {@link AgentFileStreamService.AgentFileResource}
+     */
+    public static AgentFileStreamService.AgentFileResource forAgentFile(
+        final URI uri,
+        final long size,
+        final Instant lastModifiedTime,
+        final Path relativePath,
+        final String jobId,
+        final InputStream inputStream
+    ) {
+        final String description = AgentFileResourceImpl.class.getSimpleName()
+            + " ["
+            + " jobId:" + jobId + ", "
+            + " relativePath: " + relativePath.toString()
+            + " ]";
+
+        final Path filenamePath = relativePath.getFileName();
+
+        if (filenamePath == null) {
+            throw new IllegalArgumentException("Invalid relative path, not a file");
+        }
+
+        return new AgentFileResourceImpl(
+            true,
+            uri,
+            size,
+            lastModifiedTime.toEpochMilli(),
+            filenamePath.toString(),
+            description,
+            inputStream
+        );
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean exists() {
+        return this.exists;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public boolean isOpen() {
+        return true;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public URL getURL() throws IOException {
+        //TODO: Doable, could resolve to API URL. Unclear what value this would add.
+        throw new IOException("Cannot resolve to a URL");
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public URI getURI() throws IOException {
+        if (!this.exists()) {
+            throw new IOException("Cannot determine URI of non-existent resource");
+        }
+        return this.uri;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public File getFile() throws IOException {
+        throw new FileNotFoundException("Resource is not a local file");
```

Comment:
Perhaps "UnsupportedOperationException" fits better

Reply:
This implementation complies to what the `Resource` interface declares.